Emit fully qualified namespaces

N-Olbert opened the pull request:
Fixes #1 

Also done: 
- Adjusted tests so that the actual output of the generator is verified using [snapshooter](https://github.com/SwissLife-OSS/snapshooter).

N-Olbert committed:
Adjust tests so that the actual output of the generator is verified using snapshooter.

N-Olbert committed:
Emit fully qualified namespaces

Fixes #1

N-Olbert committed:
Applied [StringSyntax("c#-test")]

N-Olbert commented:
Test failure seems not directly related to the changes...

```
Unhandled exception. System.TypeLoadException: Could not load type 'Microsoft.Testing.Platform.OutputDevice.FormattedTextOutputDeviceDataBuilder' from assembly 'Microsoft.Testing.Platform, Version=1.6.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
            at Microsoft.Testing.Extensions.TrxReport.Abstractions.TrxReportGenerator.OnTestSessionFinishingAsync(SessionUid sessionUid, CancellationToken cancellationToken)
```

I´ll look at this later.

AnderssonPeter commented:
[snapshooter](https://github.com/SwissLife-OSS/snapshooter) seems like a great idea for this type of project! The exception in the pipeline is new, so the question is if snapshooter is compatible with TUnit?

N-Olbert committed:
Update Nuget

N-Olbert committed:
Fix Warning CS1998

N-Olbert committed:
Adjust workflow for debugging purposes

N-Olbert committed:
Adjusted workflow

N-Olbert committed:
Revert Changes to workflow

N-Olbert committed:
**Really** revert the changes to test.yml

N-Olbert committed:
Upload test results for debugging purposes

N-Olbert committed:
Use testspace only on push

N-Olbert commented:
@AnderssonPeter , Snapshooter itself is not the problem; it has a specific TUnit package. I fixed the test run by updating the dependencies, and at least the tests were running after this.

Anyways, the pipeline failed. I verified that the reports are still being generated by uploading them as an artifact. This works, as seen here: https://github.com/AnderssonPeter/Dolly/actions/runs/14762250306.

Then I figured out that Testspace requires the upload of the report file, but this happens only within a 'push.' I fixed this too, by using testspace within a 'push' only (never used it before, so this may or may not be you intension)

TL;DR: It should work now. 👍

After this PR, I may fix collections and dictionaries next (see #4 ), but my free time will be more limited in the next few weeks, I think. Anyway, this is a very handy project that I could have used way more often if it had existed earlier.

AnderssonPeter commented:
Great work, i will be traveling for a few days, but when I'm back I'll check the pr in details and merge it!
Thanks for contributing to Dolly!

AnderssonPeter commented:
Sorry for the delay, I have looked at the PR, While I agree that [snapshooter](https://github.com/SwissLife-OSS/snapshooter) is great for testing code generation, I'm not sure I'm that fond of using it to validate code parsing, it makes it harder to test only part of the model parsed, so one change might break multiple tests even when it only should affect a few of them.

AnderssonPeter commented:
Sorry for the late reply, there are a few tests where snapshooter, might not be a good option:
* `ParseModelFlags` it only makes the test more complicated to read/validate.

One additional downside to using snapshooter this way is that we are validating the output string instead of the sate of the object, how about changing it to use json serialization (with string enums) instead of to string, that way we would compare the state and not the output.

That being said some tests that construct models and validate the output would be nice too, and those can use snapshooter :)

N-Olbert commented:
No problem, late reply here too 😄 

> how about changing it to use json serialization (with string enums) instead of to string, that way we would compare the state and not the output.

I’m happy to make that change! Just one small remark - does anyone really care about the business model in a project with such a small and well-constrained business domain as a source generator? By introducing Snapshooter, I indeed essentially converted all these tests into end-to-end tests (except for a final "compile the output" stage, which may be useful too). From my perspective, I don’t find the model important enough to be validated: the output is all that matters for the given input, and anything else is mostly irrelevant. But of course, that’s just my opinion, and as I said, I'll go ahead with the proposed change 😏 Just be aware that this may lead to more snapshots needing adjustments in the future even for simple changes - which is probably actually desirable.

To sum up: If I understood correctly, I should additionally verify the model  using snapshots instead of only verifying the output, right?

Regarding:

> `ParseModelFlags` only makes the test more complicated to read/validate.

I think this check can be removed completely, since it would be included after that change, correct?

AnderssonPeter commented:
`does anyone really care about the business model in a project with such a small and well-constrained business domain as a source generator` It's easier to spot bugs if a unit test only tests one thing instead of multiple, so while not strictly necessary it's prefeered.

About `ParseModelFlags` I'm not 100% sure, i guess it will be easier to check after the other changes have been made?

AnderssonPeter commented:
@N-Olbert Hi any chance you could fix the last issue so this can be merged?

Thanks

N-Olbert commented:
@AnderssonPeter Not before July. The core fix is just the change within `SourceTextConverter.cs` (and the pipeline adjustment in tests.yaml). If needed before, I would just recommend undoing all other changes and merge it (can't do it atm since I am on vacation 🌴 ).

N-Olbert committed:
Revert SnapHooter-migration of tests since it is not really part of the issue.

N-Olbert committed:
Adjusted tests

N-Olbert committed:
Update test.yml

N-Olbert committed:
Nit: Fix spacing

N-Olbert commented:
@AnderssonPeter This is ready. I revoked the changes regarding SnapShooter, since they aren`t strictly required / part of the issue.
